theorems/homotopy/FinSet.agda

```diff
@@ -0,0 +1,97 @@
+{-# OPTIONS --without-K --rewriting #-}
+
+open import HoTT
+import homotopy.ConstantToSetExtendsToProp as ConstExt
+open import homotopy.Pigeonhole
+
+module homotopy.FinSet where
+
+-- the explicit type of finite sets, carrying the cardinality on its sleeve
+FinSet-exp : Type₁
+FinSet-exp = Σ ℕ λ n → BAut (Fin n)
+
+FinSet-prop : SubtypeProp Type₀ (lsucc lzero)
+FinSet-prop = (λ A → Trunc -1 (Σ ℕ λ n → Fin n == A)) , λ A → Trunc-level
+
+-- the implicit type of finite sets, hiding the cardinality in prop. trunc.
+FinSet : Type₁
+FinSet = Subtype FinSet-prop
+
+FinSet= : {A B : FinSet} → fst A == fst B → A == B
+FinSet= = Subtype=-out FinSet-prop
+
+FinFS : ℕ → FinSet
+FinFS n = Fin n , [ n , idp ]
+
+UnitFS : FinSet
+UnitFS = ⊤ , [ S O , ua Fin-equiv-Coprod ∙ ap (λ C → C ⊔ ⊤) (ua Fin-equiv-Empty)
+                     ∙ ua (Coprod-unit-left ⊤) ]
+
+Fin-inj-lemma : {n m : ℕ} → n < m → Fin m == Fin n → ⊥
+Fin-inj-lemma {n} {m} n<m p = i≠j (<– (ap-equiv (coe-equiv p) i j) q)
+  where
+    i : Fin m
+    i = fst (pigeonhole n<m (coe p))
+
+    j : Fin m
+    j = fst (snd (pigeonhole n<m (coe p)))
+
+    i≠j : i ≠ j
+    i≠j = fst (snd (snd (pigeonhole n<m (coe p))))
+
+    q : coe p i == coe p j
+    q = snd (snd (snd (pigeonhole n<m (coe p))))
+    
+Fin-inj : (n m : ℕ) → Fin n == Fin m → n == m
+Fin-inj n m p with ℕ-trichotomy n m
+Fin-inj n m p | inl q = q
+Fin-inj n m p | inr (inl q) = ⊥-rec (Fin-inj-lemma q (! p))
+Fin-inj n m p | inr (inr q) = ⊥-rec (Fin-inj-lemma q p)
+
+FinSet-aux-prop : (A : Type₀) → SubtypeProp ℕ (lsucc lzero)
+FinSet-aux-prop A = (λ n → Trunc -1 (Fin n == A)) , (λ n → Trunc-level)
+  
+FinSet-aux : (A : FinSet) → Subtype (FinSet-aux-prop (fst A))
+FinSet-aux (A , tz) = CE.ext tz
+  where
+    to : (Σ ℕ λ n → Fin n == A) → Subtype (FinSet-aux-prop A)
+    to (n , p) = n , [ p ]
+
+    to-const : (z₁ z₂ : Σ ℕ λ n → Fin n == A) → to z₁ == to z₂
+    to-const (n₁ , p₁) (n₂ , p₂) = Subtype=-out (FinSet-aux-prop A) (Fin-inj n₁ n₂ (p₁ ∙ ! p₂))
+
+    module CE =
+      ConstExt (Subtype-level (FinSet-aux-prop A) ℕ-is-set) to to-const
+
+card : FinSet → ℕ
+card A = fst (FinSet-aux A)
+
+card-eq : (A : FinSet) (n : ℕ) → Trunc -1 (Fin n == fst A) → card A == n
+card-eq (A , tz) n tp = Trunc-rec (ℕ-is-set (card (A , tz)) n) (λ p →
+  ap card (FinSet= {A , tz} {FinFS n} (! p)) ) tp
+
+FinSet-equiv : FinSet-exp ≃ FinSet
+FinSet-equiv = equiv to from to-from from-to
```

Comment:
I might change this to `FinSet-equiv` later because `f-equiv` stands for `f` being an equivalence and `X-econv` is an equivalence with one side being `X`.